### PR TITLE
RSC compatibility

### PR DESCRIPTION
**Pull request by vladmoroz**
- Adds `'use client'` in key places
- Fixes a hydration issue with Portal SSR (https://github.com/radix-ui/primitives/issues/1386)

### CODE REVIEW

File: packages/react/primitive/src/Primitive.tsx [hunk 1]
```diff
@@ -46,9 +46,9 @@ const Primitive = NODES.reduce((primitive, node) => {
     const { asChild, ...primitiveProps } = props;
     const Comp: any = asChild ? Slot : node;
 
-    React.useEffect(() => {
+    if (typeof window !== 'undefined') {
       (window as any)[Symbol.for('radix-ui')] = true;
-    }, []);
+    }
```
Comment (vladmoroz) [inline packages/react/primitive/src/Primitive.tsx:51]: Note this change, this is to allow simple primitives to be server components (like `AccessibleIcon`)

No context for how it's used in the original PR, but seems benign to do the assignment in render:
https://github.com/radix-ui/primitives/pull/853

File: packages/react/portal/src/Portal.tsx [hunk 2]
```diff
@@ -16,11 +17,14 @@ interface PortalProps extends PrimitiveDivProps {
   /**
    * An optional container where the portaled content should be appended.
    */
-  container?: HTMLElement | null;
+  container?: Element | null;
```
Comment (vladmoroz) [inline packages/react/portal/src/Portal.tsx:20]: Corresponds to the type expected by `React.createPortal`

File: packages/react/portal/src/Portal.tsx [hunk 3]
```diff
@@ -16,11 +17,14 @@ interface PortalProps extends PrimitiveDivProps {
...
 }
 
 const Portal = React.forwardRef<PortalElement, PortalProps>((props, forwardedRef) => {
-  const { container = globalThis?.document?.body, ...portalProps } = props;
+  const { container: containerProp, ...portalProps } = props;
+  const [mounted, setMounted] = React.useState(false);
+  useLayoutEffect(() => setMounted(true), []);
+  const container = containerProp || (mounted && globalThis?.document?.body);
```
Comment (vladmoroz) [inline packages/react/portal/src/Portal.tsx:27]: When a container isn't provided, we must render twice to avoid hydration errors

**Review by hadihallak — APPROVED** [review]

File: packages/react/portal/src/Portal.tsx (hunk 3 above)
Comment (dubzzz) [inline packages/react/portal/src/Portal.tsx:27]: From an end-user point-of-view don't we introduce an extra frame as React will not render the content at first execution but after a second render cycle?

Comment (dubzzz) [reply packages/react/portal/src/Portal.tsx:27]: Actually users can still pass the container to `globalThis?.document?.body` to drop that extra frame. Forget about my comment

**Pull request merged**